File: SimWheelPS4/src/com/sdf/simwheelps4/MappingActivity.java
package com.sdf.simwheelps4;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;
import java.util.LinkedHashSet;
import java.util.Set;

/**
 * Lets the user remap every PS4 button to one of the receiver codes. Profiles
 * are kept in a single SharedPreferences slot (a JSON dictionary of profile-name
 * → mapping-json). The "live" profile is mirrored into Settings.mappingProfileJson
 * which is what MainActivity loads at startup.
 */
public class MappingActivity extends Activity {

    private static final String PROFILES_PREF = "simwheel_ps4_profiles";

    private Settings settings;
    private ButtonMap map;
    private ButtonMap.Choice[] choices;
    private TextView lastPress;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_mapping);

        settings = new Settings(this);
        map = new ButtonMap();
        map.fromJson(settings.mappingProfileJson);
        choices = ButtonMap.receiverChoices();

        lastPress = (TextView) findViewById(R.id.last_press);
        LinearLayout container = (LinearLayout) findViewById(R.id.mapping_container);
        LayoutInflater inf = LayoutInflater.from(this);

        for (final ButtonMap.Entry e : map.all()) {
            View row = inf.inflate(R.layout.row_mapping, container, false);
            ((TextView) row.findViewById(R.id.row_btn_name)).setText(e.name);
            Spinner sp = (Spinner) row.findViewById(R.id.row_spinner);
            ArrayAdapter<ButtonMap.Choice> ad = new ArrayAdapter<ButtonMap.Choice>(this,
                    android.R.layout.simple_spinner_item, choices);
            ad.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
            sp.setAdapter(ad);
            // preselect current
            int sel = 0;
            for (int i = 0; i < choices.length; i++) if (choices[i].code == e.receiverCode) { sel = i; break; }
            sp.setSelection(sel);
            sp.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
                @Override public void onItemSelected(AdapterView<?> p, View v, int pos, long id) {
                    e.receiverCode = choices[pos].code;
                    settings.mappingProfileJson = map.toJson();
                    settings.save();
                }
                @Override public void onNothingSelected(AdapterView<?> p) {}
            });
            container.addView(row);
        }

        findViewById(R.id.btn_save).setOnClickListener(new View.OnClickListener() {
            @Override public void onClick(View v) { promptSaveProfile(); }
        });
        findViewById(R.id.btn_load).setOnClickListener(new View.OnClickListener() {
            @Override public void onClick(View v) { promptLoadProfile(); }
        });
        findViewById(R.id.btn_reset_map).setOnClickListener(new View.OnClickListener() {
            @Override public void onClick(View v) {
                map = new ButtonMap();
                settings.mappingProfileJson = map.toJson();
                settings.save();
                recreate();
            }
        });
    }

    private SharedPreferences profiles() {
        return getApplicationContext().getSharedPreferences(PROFILES_PREF, Context.MODE_PRIVATE);
    }

    private void promptSaveProfile() {
        final EditText input = new EditText(this);
        input.setHint("Profile name (e.g. ETS2, F1)");
        new AlertDialog.Builder(this)
                .setTitle("Save profile")
                .setView(input)
                .setPositiveButton("Save", new DialogInterface.OnClickListener() {
                    @Override public void onClick(DialogInterface d, int w) {
                        String name = input.getText().toString().trim();
                        if (name.isEmpty()) return;
                        profiles().edit().putString(name, map.toJson()).apply();
                        Toast.makeText(MappingActivity.this, "Saved: " + name, Toast.LENGTH_SHORT).show();
                    }
                })
                .setNegativeButton("Cancel", null)
                .show();
    }

    private void promptLoadProfile() {
        Set<String> keys = new LinkedHashSet<String>(profiles().getAll().keySet());
        if (keys.isEmpty()) {
            Toast.makeText(this, "No saved profiles yet", Toast.LENGTH_SHORT).show();
            return;
        }
        final String[] arr = keys.toArray(new String[0]);
        new AlertDialog.Builder(this)
                .setTitle("Load profile")
                .setItems(arr, new DialogInterface.OnClickListener() {
                    @Override public void onClick(DialogInterface d, int which) {
                        String s = profiles().getString(arr[which], "");
                        map.fromJson(s);
                        settings.mappingProfileJson = map.toJson();
                        settings.save();
                        recreate();
                    }
                })
                .show();
    }

    /* Show button presses live so the user can identify which physical button maps where. */
    @Override
    public boolean dispatchKeyEvent(KeyEvent event) {
        if (event.getAction() == KeyEvent.ACTION_DOWN && ControllerInput.isFromGamepad(event.getDevice())) {
            ButtonMap.Entry e = map.forKey(event.getKeyCode());
            String name = (e != null) ? e.name : ("KeyCode " + event.getKeyCode());
            lastPress.setText("Pressed: " + name);
        }
        return super.dispatchKeyEvent(event);
    }
    @Override
    public boolean dispatchGenericMotionEvent(MotionEvent ev) {
        return super.dispatchGenericMotionEvent(ev);
    }
}

File: SimWheelPS4/src/main/kotlin/com/simwheel/ps4/ui/MappingActivity.kt
package com.simwheel.ps4.ui

import android.app.Activity
import android.app.AlertDialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import android.view.KeyEvent
import com.simwheel.ps4.R
import com.simwheel.ps4.model.ButtonMapping
import com.simwheel.ps4.model.DefaultMappings
import com.simwheel.ps4.model.Settings as AppSettings

class MappingActivity : Activity() {

    private lateinit var settings: AppSettings
    private var currentMappings: MutableList<ButtonMapping> = mutableListOf()
    private var profileNames: MutableList<String> = mutableListOf()
    private lateinit var adapter: MappingAdapter
    private lateinit var spinnerProfile: Spinner
    private var waitingForKey = false
    private var editingIndex = -1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_mapping)

        settings = AppSettings(this)

        spinnerProfile = findViewById(R.id.spinner_profile)
        val listView: ListView = findViewById(R.id.lv_mappings)

        profileNames = settings.getProfileNames().toMutableList()
        if (profileNames.isEmpty()) profileNames.add("Default")
        val profileAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, profileNames)
        profileAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spinnerProfile.adapter = profileAdapter

        val activeIdx = profileNames.indexOf(settings.getActiveProfile())
        if (activeIdx >= 0) spinnerProfile.setSelection(activeIdx)

        loadProfile(settings.getActiveProfile())

        adapter = MappingAdapter()
        listView.adapter = adapter

        listView.setOnItemClickListener { _, _, position, _ ->
            showRemapDialog(position)
        }

        spinnerProfile.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View?, pos: Int, id: Long) {
                val name = profileNames[pos]
                settings.setActiveProfile(name)
                loadProfile(name)
                adapter.notifyDataSetChanged()
            }
            override fun onNothingSelected(parent: AdapterView<*>) {}
        }

        findViewById<Button>(R.id.btn_save_profile).setOnClickListener {
            val name = profileNames[spinnerProfile.selectedItemPosition]
            settings.saveMappingProfile(name, currentMappings)
            Toast.makeText(this, "Profile \"$name\" saved", Toast.LENGTH_SHORT).show()
        }

        findViewById<Button>(R.id.btn_new_profile).setOnClickListener {
            showNewProfileDialog()
        }

        findViewById<Button>(R.id.btn_reset_mappings).setOnClickListener {
            currentMappings = DefaultMappings.ALL.toMutableList()
            adapter.notifyDataSetChanged()
        }

        findViewById<Button>(R.id.btn_back_mapping).setOnClickListener { finish() }
    }

    private fun loadProfile(name: String) {
        currentMappings = (settings.loadMappingProfile(name) ?: DefaultMappings.ALL).toMutableList()
    }

    private fun showRemapDialog(index: Int) {
        val mapping = currentMappings[index]
        val pcCodes = DefaultMappings.PC_CODE_NAMES.keys.sorted()
        val pcNames = pcCodes.map { DefaultMappings.pcCodeName(it) }

        val currentIdx = pcCodes.indexOf(mapping.pcCode).takeIf { it >= 0 } ?: 0

        AlertDialog.Builder(this)
            .setTitle("Remap: ${mapping.ps4Name}")
            .setItems(pcNames.toTypedArray()) { _, which ->
                val newCode = pcCodes[which]
                currentMappings[index] = mapping.copy(
                    pcCode = newCode,
                    pcName = DefaultMappings.pcCodeName(newCode)
                )
                adapter.notifyDataSetChanged()
            }
            .setNegativeButton("Cancel", null)
            .show()
    }

    private fun showNewProfileDialog() {
        val input = EditText(this)
        input.hint = "Profile name"
        AlertDialog.Builder(this)
            .setTitle("New Profile")
            .setView(input)
            .setPositiveButton("Create") { _, _ ->
                val name = input.text.toString().trim()
                if (name.isNotEmpty() && !profileNames.contains(name)) {
                    profileNames.add(name)
                    settings.saveMappingProfile(name, currentMappings)
                    (spinnerProfile.adapter as ArrayAdapter<*>).notifyDataSetChanged()
                    spinnerProfile.setSelection(profileNames.indexOf(name))
                }
            }
            .setNegativeButton("Cancel", null)
            .show()
    }

    inner class MappingAdapter : BaseAdapter() {
        override fun getCount() = currentMappings.size
        override fun getItem(pos: Int) = currentMappings[pos]
        override fun getItemId(pos: Int) = pos.toLong()

        override fun getView(pos: Int, convertView: View?, parent: ViewGroup): View {
            val view = convertView ?: LayoutInflater.from(this@MappingActivity)
                .inflate(R.layout.item_mapping, parent, false)
            val m = currentMappings[pos]
            view.findViewById<TextView>(R.id.tv_ps4_button).text = m.ps4Name
            view.findViewById<TextView>(R.id.tv_pc_action).text  = m.pcName
            view.findViewById<TextView>(R.id.tv_code).text       = "[${m.pcCode}]"
            return view
        }
    }
}
